### PR TITLE
feat(gomod): support -e option

### LINKED ISSUE #19816: Support go mod tidy -e Option
### What would you like Renovate to be able to do?

We are considering enabling the **postConfiguration** option **gomodTidy**, so we don't have to keep fixing broken PRs due to bad **go.sum** files.

But many of our repos are incomplete in source-repo form, so in fact we run `go mod tidy -e` so the **go.mod** and **go.sum** files are updated even in the presence of errors (of the sort that don't prevent some degree of tidying, which is what we care about).

Would it be possible to support an option to enable that?

### If you have any ideas on how this should be implemented, please tell us here.

I'm not sure what the new option should be called; maybe **goModTidyProceed**, based on the help text?

> The -e flag causes tidy to attempt to proceed despite errors encountered while loading packages.

### Is this a feature you are interested in implementing yourself?

Maybe

### PR DESCRIPTION
**Pull request by BirknerAlex**
## Changes

Implementing `go mod tidy -e` support by adding a new option `gomodTidyE`.

## Context

- Closes #19816

## Documentation

- [X] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work

I have verified these changes via:

- [ ] Code inspection only, or
- [X] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/manager/gomod/artifacts.ts
Comment (viceice) [inline lib/modules/manager/gomod/artifacts.ts:302]: shouldn't we allow both options?

Comment (BirknerAlex) [reply lib/modules/manager/gomod/artifacts.ts:302]: I can change that if you like. So if both options are present it adds both arguments. Will implement that next week. 👍

Comment (viceice) [reply lib/modules/manager/gomod/artifacts.ts:302]: yes, please do .  but only if they are allowed to be used together otherwise warn

Comment (BirknerAlex) [reply lib/modules/manager/gomod/artifacts.ts]: Done

File: docs/usage/golang.md
Comment (viceice) [inline docs/usage/golang.md:18]:
```suggestion
1. If the user has enabled the option `gomodTidy`, `gomodTidy1.17` or `gomodTidyE` in the [`postUpdateOptions`](https://docs.renovatebot.com/configuration-options/#postupdateoptions) array, then Renovate runs `go mod tidy`, `go mod tidy -compat=1.17` or `go mod tidy -e` respectively, which itself can update `go.mod` and `go.sum`.
```

**Review by viceice — COMMENTED** [review]
a small one

File: docs/usage/golang.md
Comment (HonkingGoose) [reply docs/usage/golang.md:18]:
```suggestion
1. If the user has enabled the option `gomodTidy`, `gomodTidy1.17` or `gomodTidyE` in the [`postUpdateOptions`](https://docs.renovatebot.com/configuration-options/#postupdateoptions) array, then Renovate runs `go mod tidy`, `go mod tidy -compat=1.17` or `go mod tidy -e` respectively, which itself can update `go.mod` and `go.sum`
```

We don't use punctuation at the end of list items.

File: docs/usage/golang.md
Comment (HonkingGoose) [inline docs/usage/golang.md:20]:
```suggestion
   1. If adding `gomodTidy1.17` and `gomodTidyE` at the same time, both options will be enabled and results in running `go mod tidy -compat=1.17 -e`
```

Same thing here.

File: lib/modules/manager/gomod/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/gomod/readme.md:8]:
```suggestion
1. `gomodTidy` - if you'd like Renovate to run `go mod tidy` after every update before raising the PR
   1. This is implicitly enabled for major updates if the user has enabled the option `gomodUpdateImportPaths`
1. `gomodTidy1.17` - if you'd like Renovate to run `go mod tidy -compat=1.17` after every update before raising the PR
1. `gomodTidyE` - if you'd like Renovate to run `go mod tidy -e` after every update before raising the PR
1. `gomodUpdateImportPaths` - if you'd like Renovate to update your source import paths on major updates before raising the PR
1. `gomodMassage` - to enable massaging of all `replace` statements prior to running `go` so that they will be ignored
```

Comment (BirknerAlex) [reply lib/modules/manager/gomod/readme.md]: Removed all, also on existing list elements.

File: docs/usage/golang.md
Comment (HonkingGoose) [inline docs/usage/golang.md:18]: Do we _really_ want this big list of internal details? This one bullet point now has:

Three variants of the `go mod tidy` command:

- `go mod tidy`
- `go mod tidy -compat=1.17`
- `go mod tidy -compat=1.17`

Two file formats:

- `go.mod`
- `go.sum`

So we're showing five things in one bullet point. I find this hard to understand, and it's easy to lose track of the five things while reading.

Comment (BirknerAlex) [reply docs/usage/golang.md:18]: For me personally I am always a big fan of knowing which commands are executed and not having a "black box" from end-user perspective. But it's your project and your decision.  :)

Comment (HonkingGoose) [reply docs/usage/golang.md:18]: Thank you for your thoughts on this! 😄

My point is: the current layout of the information is hard to read and understand. To summarize:

- If we decide this information is important for our users: create new issue to rewrite the section after this PR lands
- If this information is not key for our users: drop it

I'll let the maintainers decide what to do here. 😉

Comment (secustor) [reply docs/usage/golang.md:18]:
```suggestion
1. If the user has any of the available `gomodTidy` options ( e.g. `gomodTidy1.17` ) in the [`postUpdateOptions`](https://docs.renovatebot.com/configuration-options/#postupdateoptions), then Renovate runs `go mod tidy` with the respective options ( multiple options are allowed ).
```
I think the mentioning of `go.mod` and `go.sum` is not necessary if someone wants to modify options they will know Go modules and if not they will `go mod tidy`. This will direct them to more sophisticated documentation regarding this than we can provide.

**Review by secustor — CHANGES_REQUESTED** [review]

File: docs/usage/golang.md
Comment (secustor) [inline docs/usage/golang.md:20]:
```suggestion
```

**Review by secustor — CHANGES_REQUESTED** [review]

**Comment by BirknerAlex** [comment]
@secustor Thank you for the suggestions.

File: docs/usage/golang.md
Comment (HonkingGoose) [inline docs/usage/golang.md:18]:
```suggestion
1. If the user has any of the available `gomodTidy` options (e.g. `gomodTidy1.17`) in the [`postUpdateOptions`](https://docs.renovatebot.com/configuration-options/#postupdateoptions), then Renovate runs `go mod tidy` with the respective options (multiple options are allowed).
```

**Review by secustor — APPROVED** [review]

**Review by viceice — APPROVED** [review]

**Pull request merged**
